### ISSUE TITLE
appropriately sum dependent_variable_biomass_components 

### ISSUE THREAD

**Issue by teixeirak** [forestgeo/allodb#82]
In the long run, it won't be a stable solution to sum equations under the assumption that they describe (the desired set of) different biomass components. Rather, we will need to create equations describing how the dependent_variable_biomass_components relate to one another and use that as the basis for summing.

@teixeirak needs to explain this better.

**Comment by teixeirak** [comment]
@maurolepore, I'll describe the desired functionality and let you advise on how best to implement this code-wise. I think this will probably involve a small table.

As you know, there are quite a few cases where multiple equations need to be summed. Examples:
 - gower_1992_cdor :  `Total aboveground biomass= Stem biomass (without bark) + Stem bark biomass + Branches (live, dead) + Foliage total`

- means_ : `Total aboveground biomass=  Branches (live) + Foliage total + Stem biomass (with bark)`

In these cases, we want the code to check that the components sum as they should. That is, when sitespecies table directs to 3 equations, the code should check that they sum as expected (with expectations defined in a table?). 

There may also be some cases where minor components are missing and a warning message should be given. For example,  Abies concolor has an equation that's  just for `Stem biomass (with bark)`, whereas 
`Total aboveground biomass=  Branches (live) + Foliage total + Stem biomass (with bark)`.**  In this case, the code should send a warning such as "aboveground biomass calculations for Abies concolor do not include foliage and branches."

** There may be ambiguity in terms of how these terms are defined, and in some cases there may be ambiguity as to what's complete (e.g., I noticed that some multi-equation models include live and dead branches separately, whereas some just have live branches). We will need help from Erika to define the full set of such relationships that occur in the database/ describe appropriate ways to sum.

One way to make this a bit simpler may be along the lines of what you suggested in issue #84, but with the caveat that we'll still need to define relationships among variables. Perhaps when the equations are from the same study/site and are intended to be summed, they could all receive the same equationID, and we could add another field to describe the parts. This would probably be the easier way to start.

**Comment by maurolepore** [comment]
@teixeirak, thanks for the explanation. Sounds good. 

My suggestion is still the same, i.e. to record in the table the number of equations that you expect. Althoght you said you are not convinced that's not a good way, I can't think of a simpler way. But please let me know if you do think of a better alternative. 


Below is a mock implementation of what I mean. If you add a column that records the total number of equations you expect to sum, I can compare that number with the count of equations that will be actually summed and flag the problematic rows. 


``` r
library(tidyverse)

census <- tribble(
  ~rowid,   ~dbh, ~sp,
  1,        11,   "A a",
  2,        22,   "B b",
)
census
#> # A tibble: 2 x 3
#>   rowid   dbh sp   
#>   <dbl> <dbl> <chr>
#> 1     1    11 A a  
#> 2     2    22 B b

allodb <- tribble(
  ~eqn_id,  ~sp,   ~equation,  ~part,
  "aaa111", "A a", "dbh * 99", "whole",
  
  # Pretend that only two equations are expected
  "bbb111", "B b", "dbh * 10", "stem",
  "bbb222", "B b", "dbh * 10", "bark",
)
allodb
#> # A tibble: 3 x 4
#>   eqn_id sp    equation part 
#>   <chr>  <chr> <chr>    <chr>
#> 1 aaa111 A a   dbh * 99 whole
#> 2 bbb111 B b   dbh * 10 stem 
#> 3 bbb222 B b   dbh * 10 bark


# The code already does this ----------------------------------------------

# We first match equations like this
internal_match <- census %>% 
  left_join(allodb)
#> Joining, by = "sp"
internal_match
#> # A tibble: 3 x 6
#>   rowid   dbh sp    eqn_id equation part 
#>   <dbl> <dbl> <chr> <chr>  <chr>    <chr>
#> 1     1    11 A a   aaa111 dbh * 99 whole
#> 2     2    22 B b   bbb111 dbh * 10 stem 
#> 3     2    22 B b   bbb222 dbh * 10 bark

# We then compute biomass for each rowid
biomass <- internal_match %>% 
  mutate(biomass = eval(parse(text = .data$equation), envir = list(dbh)))
biomass
#> # A tibble: 3 x 7
#>   rowid   dbh sp    eqn_id equation part  biomass
#>   <dbl> <dbl> <chr> <chr>  <chr>    <chr>   <dbl>
#> 1     1    11 A a   aaa111 dbh * 99 whole     110
#> 2     2    22 B b   bbb111 dbh * 10 stem      220
#> 3     2    22 B b   bbb222 dbh * 10 bark      220

# We finally group by rowid (not by equation id) and sum all the equations.
# BUT WE HAVE NO WAY TO KNOW IF ALL THE EXPECTED EQUATIONS ARE INCLUDED!
biomass %>% 
  group_by(rowid) %>% 
  summarise(final_biomass = sum(biomass))
#> # A tibble: 2 x 2
#>   rowid final_biomass
#>   <dbl>         <dbl>
#> 1     1           110
#> 2     2           440



# Add the total number of expected equations ------------------------------

# Compare:

# Case when the total number of equations is as expected
allodb2 <- mutate(allodb, n_parts = c(1L, 2L, 2L))
census %>% 
  left_join(allodb2) %>% 
  mutate(biomass = eval(parse(text = .data$equation), envir = list(dbh))) %>% 
  add_count(rowid) %>% 
  group_by(rowid) %>% 
  mutate(
    warn = !identical(
      unique(n_parts), unique(n)
    )
  )
#> Joining, by = "sp"
#> # A tibble: 3 x 10
#> # Groups:   rowid [2]
#>   rowid   dbh sp    eqn_id equation part  n_parts biomass     n warn 
#>   <dbl> <dbl> <chr> <chr>  <chr>    <chr>   <int>   <dbl> <int> <lgl>
#> 1     1    11 A a   aaa111 dbh * 99 whole       1     110     1 FALSE
#> 2     2    22 B b   bbb111 dbh * 10 stem        2     220     2 FALSE
#> 3     2    22 B b   bbb222 dbh * 10 bark        2     220     2 FALSE

# Whase when the number of equations is NOT as expected
# (flag the row to later throw a warning)
allodb2 <- mutate(allodb, n_parts = c(1L, 3L, 3L))
biomass2 <- census %>% 
  left_join(allodb2) %>% 
  mutate(biomass = eval(parse(text = .data$equation), envir = list(dbh)))
#> Joining, by = "sp"

# WE NOW HAVE A WAY TO CHECK IF THE NUMBER OF EQUATIONS IS THE ONE WE EXPECT

biomass2 %>% 
  add_count(rowid) %>% 
  group_by(rowid) %>% 
  mutate(
    warn = !identical(
      unique(n_parts), unique(n)
    )
  )
#> # A tibble: 3 x 10
#> # Groups:   rowid [2]
#>   rowid   dbh sp    eqn_id equation part  n_parts biomass     n warn 
#>   <dbl> <dbl> <chr> <chr>  <chr>    <chr>   <int>   <dbl> <int> <lgl>
#> 1     1    11 A a   aaa111 dbh * 99 whole       1     110     1 FALSE
#> 2     2    22 B b   bbb111 dbh * 10 stem        3     220     2 TRUE 
#> 3     2    22 B b   bbb222 dbh * 10 bark        3     220     2 TRUE

# ... The next step is to sum biomass by rowid as I showed above and throw a warning
# pointing to the problems with the rows where `warn = TRUE`.
```

Created on 2019-03-26 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)

**Comment by maurolepore** [comment]
RE:

> One way to make this a bit simpler may be along the lines of what you suggested in issue #84, but with the caveat that we'll still need to define relationships among variables.

I don't understand what you mean by "define relationships among variables." but I guess this will become clearer once you do it.

>  Perhaps when the equations are from the same study/site and are intended to be summed, they could all receive the same equationID, and we could add another field to describe the parts.

The code already knows how to sum equations. What the code doesn't konw is how many equaitons you expect. If you record that, then can simply compare the number you provide with the count of rows with a given `rowid` (i.e. each row in the census data). See the example above.

**Comment by teixeirak** [comment]
We could do this as a temporary fix, but perhaps there's some better way that would be easy to implement.  
My concern is that it would be nice to eventually (hopefully fairly soon) use the same database for characteristics other than total aboveground biomass: for example, roots or foliage. We might want to add equations for roots, for example. In this case, your proposed numbering system would become invalid. 
Here's a proposal:
1- rename equation_id in sitespecies to agb_equation_id. Then we're clear that we're talking about aboveground biomass and could add another column for other components (e.g., roots, foliage, total biomass) in the future.
2- give the same equation_id to all components from a singles study. For example, gower_1992_cdor gives 4 equations that must be summed (48557b, 697035, 9a977b, a59bd5). Currently they're treated as fully independent. Rather, what if they're only given a single ID (e.g., agb123). That single ID appears in sitespecies (rather than 4 rows for the same equation- much cleaner). Then, a separate table links equations as appropriate: agb123 = sum(48557b, 697035, 9a977b, a59bd5).

The above structure would then be flexible to include other components. In the example above, say we find a 5th equation describing roots (roots11). We expand the sitespecies table to include fields for `root_equation_id`, `foliage_equation_id`, and `totb_equation_id`, and populate it appropriately:
 `root_equation_id` = roots11
`foliage_equation_id`= a59bd5
 `totb_equation_id` = totb11, and (in linking table) totb11= sum(48557b, 697035, 9a977b, a59bd5, roots11)

Would this work?

**Comment by maurolepore** [comment]
I think we need to talk about this. There seems to be some confsusion and it is using too much of our time. 

> RE: "ample, gower_1992_cdor gives 4 equations that must be summed (48557b, 697035, 9a977b, a59bd5). Currently they're treated as fully independent". 

I may be confused but I am afraid your assumption is not correct. The code already sums multiple biomass components. For each tree in the census dataset `allo_evaluate()` returns a single biomass value, i.e. the sum of the biomass of each equation. 

Is this clear?

**Comment by teixeirak** [comment]
Yes. What I wrote was misleading. Here's what I meant: "Currently they all have a separate entry in sitespecies (as do all fully independent equations)". 
If there are other issues you can give me a call.

**Comment by maurolepore** [comment]
Thanks Krista for your call. Here I summarize the take-aways. 

* We clarified that `equation_id` is a key column (in the database sense of the word) linking `equations` and `sitespecies`. Its name can be changed but its funtion needs to remain the same, i.e. to identify uniquely each equation individually.

* We clarified that it would be nice to have column, say, `group_id` that explicitely groups the different componentes that make up the total biomass of a tree.

* This framewor certainly allows the expansion of allodb to include, say, root allometries.
